### ISSUE TITLE
Undefined index: 0

### ISSUE THREAD

**Issue by oligriffiths** [michelf/php-markdown#285]
Hi,

I get an undefined index error with the following code:
```php
$string = '**A blazingly intelligent first book of essays from the award-winning author of *Open City* and *Every Day Is for the Thief***  
   
With this collection of more than fifty pieces on politics, photography, travel, history, and literature, Teju Cole solidifies his place as one of today’s most powerful and original voices. On page after page, deploying prose dense with beauty and ideas, he finds fresh and potent ways to interpret art, people, and historical moments, taking in subjects from Virginia Woolf, Shakespeare, and W. G. Sebald to Instagram, Barack Obama, and Boko Haram. Cole brings us new considerations of James Baldwin in the age of Black Lives Matter; the African American photographer Roy DeCarava, who, forced to shoot with film calibrated exclusively for white skin tones, found his way to a startling and true depiction of black subjects; and (in an essay that inspired both praise and pushback when it first appeared) the White Savior Industrial Complex, the system by which African nations are sentimentally aided by an America “developed on pillage.”  
   
Persuasive and provocative, erudite yet accessible, *Known and Strange Things* is an opportunity to live within Teju Cole’s wide-ranging enthusiasms, curiosities, and passions, and a chance to see the world in surprising and affecting new frames.  
   
**Praise for *Known and Strange Things***  
    
“On every level of engagement and critique, *Known and Strange Things* is an essential and scintillating journey.”**—Claudia Rankine, *The New York Times Book Review *(Editors’ Choice)**  
** **  
“Cole has fulfilled the dazzling promise of his novels *Every Day Is for the Thief* and *Open City*. He ranges over his interests with voracious keenness, laser-sharp prose, an open heart and a clear eye.”**—*The Guardian***  
*** ***  
“We have in Cole, a Nigerian American, a continuation of [James] Baldwin’s legacy; he’s an observer and truth-seeker of the highest order. . . . It is a joy to go inside the mind of someone for whom clever insight is second nature.”**—*The Seattle Times***  
*** ***  
“[A] dazzlingly wide-ranging collection.”**—*San Francisco Chronicle***  
*** ***  
“[Cole] brings a subtle, layered perspective to all he encounters—whether it’s photographs, books, foreign countries, or Internet memes. The collected essays of *Known and Strange Things* offer a glimpse of a roving mind in action.”**—*Vanity Fair***  
*** ***  
“Essays pulse with the possible; the best ones gesture at unexplored territories. But they feel most satisfying where the author has followed his ideas to places the reader hadn’t thought to visit. *Known and Strange Things* contains many essays that do this beautifully, combining the thoughtful pause with insistent questioning, tumbling over different terrains, picking up bits of them as they go, taking on the grain and texture of all the places they’ve been.”**—*Financial Times***  
*** ***  
“There’s almost no subject Cole can’t come at from a startling angle. . . . His [is a] prickly, eclectic, roaming mind.”**—*The Boston Globe***  
*** ***  
“Bold, thoughtful essays . . . Cole’s latest book feels like an intimate conversation with an eccentric friend who cannot wait to share his wonderment with the visual world. Like a modern-day Montaigne, Cole patiently teases out deeper meanings from varied art forms and the outer margins of everyday existence.”**—Minneapolis *Star Tribune***';

\Michelf\MarkdownExtra::defaultTransform($string);
```

Any ideas? The code looks fine to me.

**Comment by michelf** [comment]
I have no idea. Running that code works fine for me. Pasting this input text to [the dingus] gives a decent result.

[the dingus]: https://michelf.ca/projects/php-markdown/dingus/

**Comment by oligriffiths** [comment]
Hmm, so I wonder if this is an encoding issue. 
When I copy the value using XDebug, and manually run it though, I get 2 errors on the following lines:
https://github.com/michelf/php-markdown/blob/lib/Michelf/Markdown.php#L1327
https://github.com/michelf/php-markdown/blob/lib/Michelf/Markdown.php#L1368 

PHP error:  Undefined offset: 0 in vendor/michelf/php-markdown/Michelf/Markdown.php on line 1368
PHP error:  Undefined offset: 0 in vendor/michelf/php-markdown/Michelf/Markdown.php on line 1327

When I paste the value in a gist, and then copy it back from the `raw` format, and re-run that, it works fine without errors.

**Comment by oligriffiths** [comment]
Here, try this file content:
[test.txt](https://github.com/michelf/php-markdown/files/1373329/test.txt)

**Comment by michelf** [comment]
Ah, I see now. Somehow, the garden of asterisks is confusing the emphasis parser. More specifically, the asterisk following "The New York Times Book Review" is not a valid emphasis marker (wrong side of the space), this debalances the markers and causes the parser to enter into an inconsistent state. It then emits PHP notices later in the process. It took me a few hours to solve that puzzle.

The difference between copy-pasting from a web page and using directly the text file is a few space characters and no-break spaces tend to disappear. The parser still enters its inconsistent state but apparently does not trigger PHP notices in that case.

**Comment by michelf** [comment]
Fixed in f7bc32eefb7f1ac446d94fb057e9da24d6176e67.

**Comment by oligriffiths** [comment]
Awesome. Thanks for the swift response. Very much appreciated.
